### PR TITLE
Unpersist existing Crafting Entry REs

### PR DESCRIPTION
**Pull request by ebuckle**
Finish work started in previous PR to migrate old crafting entries and related REs.

### CODE REVIEW

File: src/module/migration/migrations/773-unpersist-crafting-entries-and-add-predicate.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                    });
+                }
+            }
+            const craftingData: MaybeWithOldEntries = actorData.system.crafting;
+            craftingData["-=entries"] = null;
+        }
+    }
+
+    override async updateItem(itemSource: ItemSourcePF2e): Promise<void> {
+        itemSource.system.rules ??= [];
```
Comment (stwlam) [inline src/module/migration/migrations/773-unpersist-crafting-entries-and-add-predicate.ts:47]:
```suggestion
```
Were this actually a problem, we'd have a way more serious mess to clean up

File: src/module/migration/migrations/773-unpersist-crafting-entries-and-add-predicate.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        const rules = itemSource.system.rules;
+
+        // Change requiredTraits property to craftableItems predicate
+        const craftingEntryRules = itemSource.system.rules.filter(
+            (rule: Record<string, unknown>): rule is CraftingEntryRuleSourceOld =>
+                typeof rule.key === "string" &&
+                ["CraftingEntry", "PF2E.RuleElement.CraftingEntry"].includes(rule.key) &&
+                Array.isArray(rule["requiredTraits"])
```
Comment (stwlam) [inline src/module/migration/migrations/773-unpersist-crafting-entries-and-add-predicate.ts:55]:
```suggestion
                rule.key === "CraftingEntry" && Array.isArray(rule.requiredTraits)
```

File: src/module/migration/migrations/773-unpersist-crafting-entries-and-add-predicate.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                            .map((preparedFormula, index) => {
+                                if (formula.uuid !== preparedFormula.itemUUID) return {};
+                                return {
+                                    quantity: preparedFormula.quantity,
+                                    expended: preparedFormula.expended,
+                                    slot: index,
+                                };
+                            })
+                            .filter((s) => s.slot !== null);
+                    });
```
Comment (stwlam) [inline src/module/migration/migrations/773-unpersist-crafting-entries-and-add-predicate.ts:38]: A map that requires nothing is a for-loop.

File: src/module/migration/migrations/773-unpersist-crafting-entries-and-add-predicate.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            delete craftingEntryRule.requiredTraits;
+        }
+
+        // Add "phase":"afterDerived" property to any AE-Likes targeting "system.crafting.entries"
+        const craftingEntryAELikes = itemSource.system.rules.filter((r: RuleElementSource & { path?: unknown }) => {
+            r.key === "ActiveEffectLike" && typeof r.path === "string" && r.path.startsWith("system.crafting.entries.");
+        });
+        craftingEntryAELikes.map((r: RuleElementSource & { phase?: unknown }) => {
+            r.phase = "afterDerived";
+        });
```
Comment (stwlam) [inline src/module/migration/migrations/773-unpersist-crafting-entries-and-add-predicate.ts:78]: Change to for-loop

**Review by stwlam — CHANGES_REQUESTED** [review]

**Review by stwlam — CHANGES_REQUESTED** [review]
Fix PR title to be in imperative mood.

**Comment by ebuckle** [comment]
Changes in

**Comment by stwlam** [comment]
Does this change anything in compendium JSON?

**Review by stwlam — APPROVED** [review]

**Pull request merged**